proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+
+```astro
+---
+// Valid – constant string literal
+export const output = 'server'
+// Invalid – cannot be determined at compile-time
+export const output = static ? 'static' : 'server';
+export const output = value;
+export const output = 'ser' + 'ver';
+export let output = undefined; output = 'server';
+---
+```
+
+## Internal API
+
+Astro needs to detect/understand the `export const output` syntax. This should be implemented as a specific post-processing Vite plugin. The plugin will detect the existence of valid ESM exports of `output` (`export const output`, `export let output`, `export var output`, `export { output }`, etc) and attach the statically detected `output` value to the module metadata under the `astro` namespace. This will allow us to track the `output` mode of any public route.
+
+> **Note**
+> Astro Routes include `.astro` pages, but also API endpoints (`src/pages/api/user.ts`)! It's important that this RFC handles both of these constraints in a uniform way, hence the use of a post-processing Vite plugin rather than implementing this feature in `@astrojs/compiler`.
+
+The **build** API needs to be updated to handle the new `'hybrid'` output mode. Implementation-wise, this will be similar to the existing `'server'` implementation, but it will also adopt portions of the `'static'` build to statically generate any routes that are not handled at request time.
+
+The **dev** server must be updated to allow `searchParams` and other features which are currently only avaliable with `server` rendering to be allowed on a per-route basis.
+
+## Adapter Support
+
+An adapter will be required when building for `hybrid` output, because we will be building a server with some statically generated routes. Adapters will need to be updated to handle `hybrid` output.
+
+# Testing strategy
+
+This implementation will largely be tested with full fixture-based integration tests, because we will need to verify the build output is structured correctly, which will be difficult to do with unit tests.
+
+The Vite plugin that detects `export const output` will be unit tested against all valid `export` formats outlined above, as well as invalid dynamic exports.
+
+# Drawbacks
+
+- A new output mode will increase the complexity and maintenance costs of our `build` and `adapter` implementations.
+- The `export const output` syntax is JS-like, but since it requires static analysis, it does not support the full dynamicism of JS that users might expect. Clear errors will help ease this problem.
+- Having a separate mode for `hybrid` increases cognitive overhead and the learning curve of Astro.
+- `hybrid` output should arguably be the default `output` mode (similar to Next.js), but to avoid a breaking change we're going with an opt-in adoption approach
+
+# Alternatives
```

Comment:
Might be nice to add cache headers as an alternative here, even if just to show how many hoops have to be jumped through to properly handle caching in a CDN via headers, it's not the easiest balancing act 😅

Reply:
I completely agree that Astro should just always be hybrid!
In `astro.config` the `output` option could be renamed to `defaultOutput` to set the default for the hybrid mode. The default mode can be overridden by exporting the `output` from the frontmatter. 🤔

This would be the most flexible approach imho and also would be super logical and not change the learning curve.

- I am totally unclear how the output would look like and how we would handle adapters as we always need an ssr-adapter to do anything on the server. E.g. we would always export a directory with client and server folders, which is weird if there is no ssr happening.
- I understand that this would mean a breaking change, but still worth a though.